src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -50494,6 +50499,64 @@ static void G(object? x)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "z").WithLocation(9, 45));
         }
 
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LocalFunction_OnlyCalledInLambda_01()
+        {
+            var source = @"
+using System.Threading.Tasks;
+using System;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var foo = GetNotNullObject();
+
+        Task.Run(() => Bar());
+        Task.Run(() => Baz());
+
+        void Bar() => VerifyNotNull(foo);
+        void Baz() => Console.WriteLine(foo.ToString());
+    }
+
+    private static object GetNotNullObject() => new object();
+
+    private static bool VerifyNotNull(object instance) => instance != null;
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(44411, "https://github.com/dotnet/roslyn/issues/44411")]
+        public void LocalFunction_OnlyCalledInLambda_02()
+        {
+            var source = @"
+using System.Threading.Tasks;
+using System;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var foo = GetNotNullObject();
+
+        void Bar() => VerifyNotNull(foo);
+        void Baz() => Console.WriteLine(foo.ToString());
+
+        Task.Run(() => Bar());
+        Task.Run(() => Baz());
+    }
+
+    private static object GetNotNullObject() => new object();
+
+    private static bool VerifyNotNull(object instance) => instance != null;
+}
+";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+        }
```

Comment:
Extra testing I think we need here:

- Tests which verify that null state changes in a lambda do not persist after the lambda is evaluated
- Tests which verify that null state changes in local functions do not persist after the local function is evaluated
- Tests which verify the null state of local functions is the union of all the call sites. Essentially have a variable which is `maybe-null` in one call and `non-null` in the other and verify with in that it's `maybe-null`.  This should be tested inside a lambda as well as outside. Possible the latter test already exists.

Reply:
It is likely that these tests already exist, I can try to identify them and follow up if this is the case.

Reply:
I agree it's likely they do but want to ensure we have them given that this is the crux of the problem we've been discussing. If we can identify them I think we should annotate them with a reference to this GitHub issue so that it's easy for future developers to see the relationship. If not then ... well there isn't too much harm in having the same test twice with a different name :smile:

Reply:
We should also make sure that we have tests for anonymous methods ... cause someone is going to use them with nullable.

Reply:
I tried identifying tests for these scenarios by simply breaking the relevant part of the NullableWalker and reading the failed test output. For all 3 I didn't feel like I found a test which succinctly demonstrated each scenario, so I am adding the tests.